File: NCalculatorLib/Exprs/UnitExpr.cs
using Null.Calculator;

#nullable enable

namespace NCalculatorLib.Exprs
{
    /// <summary>
    /// basic unit
    /// 
    /// | num
    /// | fun_expr
    /// | quote_expr
    /// </summary>
    public class UnitExpr : Expr
    {
        public readonly bool Signed;
        public readonly Expr Value;

        public UnitExpr(Expr value, bool signed)
        {
            Value = value;
            Signed = signed;
        }

        public override IEnumerable<Token> EnumTokens()
        {
            if (Signed)
                yield return new Token(TokenKind.Sub, null);

            foreach (Token token in Value.EnumTokens())
                yield return token;
        }

        public override double Eval()
        {
            return Signed ? -Value.Eval() : Value.Eval();
        }
    }
}

File: NCalculatorCppLib/Exprs/unit_expr.cpp
#include "UnitExpr.h"

using namespace NCalculator;

namespace NCalculatorLibExprs
{

	UnitExpr::UnitExpr(expr *value, bool signed_Keyword) : Signed(signed_Keyword), Value(value)
	{
	}

	std::vector<token> UnitExpr::enum_tokens()
	{
		std::vector<token> vec = std::vector<token>();
		if (Signed)
		{
			vec.push_back(token(token_kind::Sub, L""));
		}

		for (auto token : Value->enum_tokens())
		{
			vec.push_back(token);
		}

		return vec;
	}

	double UnitExpr::eval()
	{
		return Signed ? -Value->eval() : Value->eval();
	}
}
